program-runtime/src/executor_cache.rs

```diff
@@ -27,81 +27,95 @@ pub trait Executor: Debug + Send + Sync {
-    /// Executor was missing from the cache, but not updated
-    Missing,
-    /// Executor is for an updated program
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum LocalGlobalDifference {
+    /// The LocalExecutorCacheEntry did not change and is the same as the GlobalExecutorCacheEntry.
+    None,
+    /// The LocalExecutorCacheEntry was inserted, no matching GlobalExecutorCacheEntry exists, so it needs to be inserted.
+    Inserted,
+    /// The LocalExecutorCacheEntry was replaced, the matching GlobalExecutorCacheEntry needs to be updated.
     Updated,
 }
 
-/// Tracks whether a given executor is "dirty" and needs to updated in the
-/// executors cache
+/// An entry of the LocalExecutorCache
 #[derive(Debug)]
-pub struct TransactionExecutor {
-    pub(crate) executor: Arc<dyn Executor>,
-    status: TransactionExecutorStatus,
+pub struct LocalExecutorCacheEntry {
+    executor: Arc<dyn Executor>,
+    difference: LocalGlobalDifference,
 }
 
-impl TransactionExecutor {
-    /// Wraps an executor and tracks that it doesn't need to be updated in the
-    /// executors cache.
-    pub fn new_cached(executor: Arc<dyn Executor>) -> Self {
-        Self {
-            executor,
-            status: TransactionExecutorStatus::Cached,
-        }
-    }
-
-    /// Wraps an executor and tracks that it needs to be updated in the
-    /// executors cache.
-    pub fn new_miss(executor: Arc<dyn Executor>) -> Self {
-        Self {
-            executor,
-            status: TransactionExecutorStatus::Missing,
-        }
-    }
+/// A subset of the GlobalExecutorCache containing only the executors relevant to one transaction
+#[derive(Default, Debug)]
+pub struct LocalExecutorCache {
+    pub executors: HashMap<Pubkey, LocalExecutorCacheEntry>,
+}
 
-    /// Wraps an executor and tracks that it needs to be updated in the
-    /// executors cache only if the transaction succeeded.
-    pub fn new_updated(executor: Arc<dyn Executor>) -> Self {
+impl LocalExecutorCache {
+    pub fn new(executable_keys: impl Iterator<Item = (Pubkey, Arc<dyn Executor>)>) -> Self {
         Self {
-            executor,
-            status: TransactionExecutorStatus::Updated,
+            executors: executable_keys
+                .map(|(key, executor)| {
+                    let entry = LocalExecutorCacheEntry {
+                        executor,
+                        difference: LocalGlobalDifference::None,
+                    };
+                    (key, entry)
+                })
+                .collect(),
         }
     }
 
-    pub fn is_missing(&self) -> bool {
-        self.status == TransactionExecutorStatus::Missing
+    pub fn get(&self, key: &Pubkey) -> Option<Arc<dyn Executor>> {
+        self.executors.get(key).map(|entry| entry.executor.clone())
     }
 
-    pub fn is_updated(&self) -> bool {
-        self.status == TransactionExecutorStatus::Updated
+    pub fn set(&mut self, key: Pubkey, executor: Arc<dyn Executor>, replacement: bool) {
```

Comment:
instead of having `set(..., true)` mean update and `set(..., false)` mean insert, wouldn't it be better to have explicit `insert` and `update`?

Also the code as it is now, will silently work if one does "insert" and it ends up being an update, or does "update" and it ends up being an insert. Those conditions should probably be asserted?

Reply:
An `Update` must not be marked `Insert`, the other way around is just performance wasted but does not cause any divergent behavior. The reason I refactored `replacement` as a parameter is because we could probably automatically infer it from the presence in the cache. But I did not want to do that in this PR.

Reply:
But does the caller have that knowledge? Does it know when it _must_ be inserting and when it _must_ be updating? Because if the caller does but the API "just works", then that information is lost and the caller has no way of asserting its invariants. The API is in a way lossy and more limiting - if I _do_ want to assert, I can't.

In general I think that APIs that are hard to misuse - ideally through static checks, but failing that through runtime checks - are preferable to APIs that "just work". If I'm reading or refactoring code months after I've written it, I want the code to tell me what the possible states are so to minimize the amount of stuff I need to keep in my brain.

For this specific case I'd look at the rust std collection API: it uses `insert` for both insert and update, and via the return value you can assert whether a previous value should have existed or not. We should probably do the same here?

(You said you're planning to follow up to this PR so it's fine by me if you want to land this as is and continue separately - this is still clearly an improvement over the previous API.)